code/framework/java-server/src/main/java/io/github/ibuildthecloud/gdapi/id/IdFormatterUtils.java

```diff
@@ -10,10 +12,60 @@
 
 public class IdFormatterUtils {
 
-    public static Object formatReference(Field field, IdFormatter formatter, Object value) {
-        return formatReference(field.getTypeEnum(), field.getSubTypeEnums(), field.getSubTypes(), formatter, value);
+    public static Object formatReference(Field field, IdFormatter formatter, Object value, SchemaFactory schemaFactory) {
+        return formatReference(field, field.getSubTypeEnums(), field.getSubTypes(), formatter, value, schemaFactory);
     }
 
+    private static Object formatReference(Field field, List<FieldType> subTypeEnums, List<String> subTypes, IdFormatter formatter, Object value, SchemaFactory schemaFactory) {
```

Comment:
Can you combine this method with the similar one below?  Seems like the only difference is that the TYPE check needs field and schemaFactory, so in the condition of TYPE just check `if (field == null || schemaFactory == null) { return value; }`